File: backend/src/models/service.model.ts
import mongoose from "mongoose";

const serviceSchema = new mongoose.Schema(
  {
    title: {
      type: String,
      required: true,
    },

    description: {
      type: String,
      required: true,
    },

    price: {
      type: Number,
      required: true,
    },

    duration: {
      type: Number,
      required: true,
    },

    category: {
      type: String,
      required: true,
    },

    image: {
      type: String,
      required: true,
    },

    capacityPerSlot: {
      type: Number,
      required: true,
    },

    slots: [
      {
        date: {
          type: String,
        },

        startTime: {
          type: String,
        },

        endTime: {
          type: String,
        },

        remainingCapacity: {
          type: Number,
        },
      },
    ],
  },

  {
    timestamps: true,
  },
);

export default mongoose.model("Service", serviceSchema);

File: frontend/lib/data/models/service_model.dart
import 'slot_model.dart';

class ServiceModel {
  final String id;
  final String title;
  final String description;
  final double price;
  final int duration;
  final String category;
  final String image;
  final int capacityPerSlot;
  final List<SlotModel> slots;

  ServiceModel({
    required this.id,
    required this.title,
    required this.description,
    required this.price,
    required this.duration,
    required this.category,
    required this.image,
    required this.capacityPerSlot,
    required this.slots,
  });

  factory ServiceModel.fromJson(Map<String, dynamic> json) => ServiceModel(
        id: json['_id'] ?? '',
        title: json['title'] ?? '',
        description: json['description'] ?? '',
        price: (json['price'] ?? 0).toDouble(),
        duration: json['duration'] ?? 60,
        category: json['category'] ?? '',
        image: json['image'] ?? '',
        capacityPerSlot: json['capacityPerSlot'] ?? 1,
        slots: (json['slots'] as List<dynamic>? ?? [])
            .map((s) => SlotModel.fromJson(s))
            .toList(),
      );

  /// Unique dates that have at least one available slot
  List<String> get availableDates {
    final seen = <String>{};
    return slots
        .where((s) => s.isAvailable && seen.add(s.date))
        .map((s) => s.date)
        .toList();
  }

  List<SlotModel> slotsForDate(String date) =>
      slots.where((s) => s.date == date).toList();
}
